### ISSUE TITLE
Another unconditional_recursion false positive in PartialEq field comparison

### ISSUE THREAD

**Issue by faern** [rust-lang/rust-clippy#12118]
### Summary

I hit another unconditional_recursion false positive. Since the fix for #12052 did not solve my issue, this seems to be a different bug?

My `PartialEq` impl computes a comparison value from `self` and `other` and compare them, but this triggers the lint. Clippy seems to think that comparing the computed values will recurse into the same `PartiaEq` impl *again*, but the computed value is of another type, so it will not.

The false positive goes away if I compare my two values with `self_value.eq(&other_value)` instead of `self_value == other_value`

### Lint Name

unconditional_recursion

### Reproducer

I tried this code:

```rust
#[derive(Copy, Clone)]
pub struct Data(());

fn compute_value(_v: Data) -> u8 {
    unimplemented!()
}

impl PartialEq for Data {
    fn eq(&self, other: &Self) -> bool {
        let self_value = compute_value(*self);
        let other_value = compute_value(*other);
        self_value == other_value
    }
}

```

I saw this happen:

```
warning: function cannot return without recursing
  --> src/lib.rs:9:5
   |
9  | /     fn eq(&self, other: &Self) -> bool {
10 | |         let self_value = compute_value(*self);
11 | |         let other_value = compute_value(*other);
12 | |         self_value == other_value
13 | |     }
   | |_____^
   |
note: recursive call site
  --> src/lib.rs:12:9
   |
12 |         self_value == other_value
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unconditional_recursion
   = note: `#[warn(clippy::unconditional_recursion)]` on by default

warning: `playground` (lib) generated 1 warning
```

I expected to see this happen:

No lint warnings

### Version

```text
rustc 1.77.0-nightly (190f4c961 2024-01-09)
binary: rustc
commit-hash: 190f4c96116a3b59b7de4881cfec544be0246d84
commit-date: 2024-01-09
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

**Comment by faern** [comment]
Ping @GuillaumeGomez

**Comment by y21** [comment]
I can't reproduce this on master. This also looks very similar to what the PR fixed and I don't think the fix for this is on nightly yet? The last sync with upstream rust was 28 Dec and the PR was merged after that

**Comment by flip1995** [comment]
See https://github.com/rust-lang/rust-clippy/pull/12062#issuecomment-1885348839: The fix will be in Friday's nighlty
